Tests/NetworkServiceTests/NetworkServiceTests.swift

```diff
@@ -1,8 +1,170 @@
-import NetworkService
 import XCTest
+import NetworkService
 
 final class NetworkServiceTests: XCTestCase {
-    func testUseMethodExistsInAPI() {
-        NetworkService().testUse()
+    
+    var networkService: NetworkService<MockEndpoint>!
+    
+    override func setUpWithError() throws {
+        networkService = NetworkService()
+        URLProtocol.registerClass(MockURLProtocol.self)
+    }
+    
+    override func tearDownWithError() throws {
+        URLProtocol.unregisterClass(MockURLProtocol.self)
+        networkService = nil
+    }
+    
+    func testSuccessfulRequestReturnsExpectedData() async throws {
+        // Given
+        let expectedData = try XCTUnwrap("Mocked Response Data".data(using: .utf8))
+        MockURLProtocol.requestHandler = { request in
+            let response = try XCTUnwrap(HTTPURLResponse(url: XCTUnwrap(request.url), statusCode: 200, httpVersion: nil, headerFields: nil))
+            return (response, expectedData)
+        }
+        
+        let endpoint = MockEndpoint()
+        
+        // When
+        let data = try await networkService.request(endpoint)
+        
+        // Then
+        XCTAssertEqual(data, expectedData)
+    }
+    
+    func testBadRequestStatusCodeInResponseTriggersBadRequestError() async throws {
+        // Given
+        MockURLProtocol.requestHandler = { request in
+            let response = try XCTUnwrap(HTTPURLResponse(url: XCTUnwrap(request.url), statusCode: 400, httpVersion: nil, headerFields: nil))
+            return (response, nil)
+        }
+        
+        let endpoint = MockEndpoint()
+        
+        // When
+        do {
+            _ = try await networkService.request(endpoint)
+            XCTFail("Expected failure, but got success")
+        } catch let error as NetworkError {
+            XCTAssertEqual(error, .badRequest)
+        } catch {
+            XCTFail("Expected NetworkError.badRequest, but got \(error)")
+        }
```

Comment:
- If we add private variable `endpoint`, then we don't need `let endpoint = MockEndpoint()`
- `XCTFail("Unexpected error: \(error)")` is less specific and lets not to rewrite the text for each test case and therefore not to forget to change it. If you agree, update the rest of the code.

```suggestion
        // Given // When
        MockURLProtocol.requestHandler = { request in
            let response = try XCTUnwrap(HTTPURLResponse(url: XCTUnwrap(request.url), statusCode: 400, httpVersion: nil, headerFields: nil))
            return (response, nil)
        }
         
        // Then
        do {
            _ = try await networkService.request(endpoint)
            XCTFail("Expected failure, but got success")
        } catch let error as NetworkError {
            XCTAssertEqual(error, .badRequest)
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
```

Reply:
Updated with 036eb7f